test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,300 @@
+				Name:   cluster.Name,
+				Images: input.E2EConfig.Images,
+			})).To(Succeed())
+		}
+
+		// Get a ClusterBroker so we can interact with the workload cluster
+		managementClusterProxy = input.BootstrapClusterProxy.GetWorkloadCluster(ctx, cluster.Namespace, cluster.Name)
+
+		// Download the v1alpha3 clusterctl version to be used for setting up the management cluster to be upgraded
+		clusterctlBinaryURL := input.E2EConfig.GetVariable(initWithBinaryVariableName)
+		clusterctlBinaryURL = strings.ReplaceAll(clusterctlBinaryURL, "{OS}", runtime.GOOS)
+		clusterctlBinaryURL = strings.ReplaceAll(clusterctlBinaryURL, "{ARCH}", runtime.GOARCH)
+
+		log.Logf("downloading clusterctl binary from %s", clusterctlBinaryURL)
+		binaryFileclusterctlBinaryPath := downloadToTmpFile(clusterctlBinaryURL)
+		defer os.Remove(binaryFileclusterctlBinaryPath) // clean up
+
+		err := os.Chmod(binaryFileclusterctlBinaryPath, 0744)
+		Expect(err).ToNot(HaveOccurred(), "failed to chmod temporary file")
+
+		By("Initializing the workload cluster with older versions of providers")
+		clusterctl.InitManagementClusterAndWatchControllerLogs(ctx, clusterctl.InitManagementClusterAndWatchControllerLogsInput{
+			ClusterctlBinaryPath:    binaryFileclusterctlBinaryPath, // use older version of clusterctl to init the management cluster
+			ClusterProxy:            managementClusterProxy,
+			ClusterctlConfigPath:    input.ClusterctlConfigPath,
+			CoreProvider:            input.E2EConfig.GetProvidersWithOldestVersion(config.ClusterAPIProviderName)[0],
+			BootstrapProviders:      input.E2EConfig.GetProvidersWithOldestVersion(config.KubeadmBootstrapProviderName),
+			ControlPlaneProviders:   input.E2EConfig.GetProvidersWithOldestVersion(config.KubeadmControlPlaneProviderName),
+			InfrastructureProviders: input.E2EConfig.GetProvidersWithOldestVersion(input.E2EConfig.InfrastructureProviders()...),
+			LogFolder:               filepath.Join(input.ArtifactFolder, "clusters", cluster.Name),
+		}, input.E2EConfig.GetIntervals(specName, "wait-controllers")...)
+
+		By("THE MANAGEMENT CLUSTER WITH THE OLDER VERSION OF PROVIDERS IS UP&RUNNING!")
+
+		Byf("Creating a namespace for hosting the %s test workload cluster", specName)
+		testNamespace, testCancelWatches = framework.CreateNamespaceAndWatchEvents(ctx, framework.CreateNamespaceAndWatchEventsInput{
+			Creator:   managementClusterProxy.GetClient(),
+			ClientSet: managementClusterProxy.GetClientSet(),
+			Name:      specName,
+			LogFolder: filepath.Join(input.ArtifactFolder, "clusters", "bootstrap"),
+		})
+
+		By("Creating a test workload cluster")
+
+		// NOTE: This workload cluster is used to check the old management cluster works fine.
+		// In this case ApplyClusterTemplateAndWait can't be used because this helpers is linked to last version of the API;
+		// so the are getting a template using the downloaded version of clusterctl, applying it, and wait for machines to be provisioned.
```

Comment:
```suggestion
		// In this case ApplyClusterTemplateAndWait can't be used because this helper is linked to the last version of the API;
		// so we are getting a template using the downloaded version of clusterctl, applying it, and wait for machines to be provisioned.
```